sklearn/decomposition/_nmf.py

```diff
@@ -907,6 +907,41 @@ def _fit_multiplicative_update(
     return W, H, n_iter
 
 
+@validate_params(
+    {
+        "X": ["array-like", "sparse matrix"],
+        "W": ["array-like", None],
+        "H": ["array-like", None],
+        "n_components": [Interval(Integral, 1, None, closed="left"), None],
+        "init": [
+            StrOptions({"random", "nndsvd", "nndsvda", "nndsvdar", "custom"}),
+            None,
+        ],
+        "update_H": ["boolean"],
+        "solver": [StrOptions({"mu", "cd"})],
+        "beta_loss": [
+            StrOptions({"frobenius", "kullback-leibler", "itakura-saito"}),
+            Real,
+        ],
+        "tol": [Interval(Real, 0, None, closed="left")],
+        "max_iter": [Interval(Integral, 0, None, closed="left")],
```

Comment:
`max_iter=0` will raise not a proper error:

```python
   ...: import numpy as np
   ...: X = np.array([[1,1], [2, 1], [3, 1.2], [4, 1], [5, 0.8], [6, 1]])
   ...: from sklearn.decomposition import non_negative_factorization
   ...: W, H, n_iter = non_negative_factorization(X, n_components=2,
   ...: init='random', random_state=0, max_iter=0)
   ...: 
   ...: 
---------------------------------------------------------------------------
UnboundLocalError                         Traceback (most recent call last)
<ipython-input-3-a85087cc471b> in <module>
      4 X = np.array([[1,1], [2, 1], [3, 1.2], [4, 1], [5, 0.8], [6, 1]])
      5 from sklearn.decomposition import non_negative_factorization
----> 6 W, H, n_iter = non_negative_factorization(X, n_components=2,
      7 init='random', random_state=0, max_iter=0)
      8 

~/Documents/packages/scikit-learn/sklearn/utils/_param_validation.py in wrapper(*args, **kwargs)
    158                 parameter_constraints, params, caller_name=func.__qualname__
    159             )
--> 160             return func(*args, **kwargs)
    161 
    162         return wrapper

~/Documents/packages/scikit-learn/sklearn/decomposition/_nmf.py in non_negative_factorization(X, W, H, n_components, init, update_H, solver, beta_loss, tol, max_iter, alpha, alpha_W, alpha_H, l1_ratio, regularization, random_state, verbose, shuffle)
   1186 
   1187     with config_context(assume_finite=True):
-> 1188         W, H, n_iter = est._fit_transform(X, W=W, H=H, update_H=update_H)
   1189 
   1190     return W, H, n_iter

~/Documents/packages/scikit-learn/sklearn/decomposition/_nmf.py in _fit_transform(self, X, y, W, H, update_H)
   1779 
   1780         if self.solver == "cd":
-> 1781             W, H, n_iter = _fit_coordinate_descent(
   1782                 X,
   1783                 W,

~/Documents/packages/scikit-learn/sklearn/decomposition/_nmf.py in _fit_coordinate_descent(X, W, H, tol, max_iter, l1_reg_W, l1_reg_H, l2_reg_W, l2_reg_H, update_H, verbose, shuffle, random_state)
    535             break
    536 
--> 537     return W, Ht.T, n_iter
    538 
    539 

UnboundLocalError: local variable 'n_iter' referenced before assignment
```

I assume that it should work. It could be worth to have a test as well.